gpflow/models/model.py

```diff
@@ -57,8 +52,30 @@ def log_prior_density(self) -> tf.Tensor:
         else:
             return to_default_float(0.0)
 
+    def log_posterior_density(self, *args, **kwargs) -> tf.Tensor:
+        """
+        This may be the posterior with respect to the hyperparameters (e.g. for
+        GPR) or the posterior with respect to the function (e.g. for GPMC and
+        SGPMC). It assumes that maximum_log_likelihood_objective() is defined
+        sensibly.
+        """
+        return self.maximum_log_likelihood_objective(*args, **kwargs) + self.log_prior_density()
+
+    def _training_loss(self, *args, **kwargs) -> tf.Tensor:
+        """
+        Training loss definition. To allow MAP (maximum a-posteriori) estimation,
+        adds the log density of all priors to maximum_log_likelihood_objective().
+        """
+        return -(self.maximum_log_likelihood_objective(*args, **kwargs) + self.log_prior_density())
```

Comment:
yeah i'm pro this - I think fewer names means users will be less doubtful about which method they should be calling, and private names gives us a chance to have some time with the new API and make any adjustments down the line if we need

Reply:
I'm sorry but I think this misses the point of the whole abstraction. The idea of the TrainingLossMixin is that all the logic relevant to training is on it. The model knows nothing about a training loss. Here you are saying, secretly the model does know about a training loss! But what happens if you want to train with different objective function, maybe you want to importance sample or something. 

Before you just implement a new Mixin and you're good to go: after all the model knows nothign about training. Now, if you do that you'll get a different `training_loss` and `training_loss_closure`  which may be completely unrelated to `_training_loss`

Reply:
```suggestion
    def maximum_log_likelihood_objective_with_priors(self, *args, **kwargs) -> tf.Tensor:
        """
        Training loss definition. To allow MAP (maximum a-posteriori) estimation,
        adds the log density of all priors to maximum_log_likelihood_objective().
        """
        return -(self.maximum_log_likelihood_objective(*args, **kwargs) + self.log_prior_density())
```

Reply:
its possible making this private is a good idea though.